README.md
# salahtimes

<p align="center" background-color="white">
<img src="https://raw.githubusercontent.com/doniseferi/salahtimes/bee6545e34e576a7467999c52c6a710815fde442/salahtimes.svg?sanitize=true" width="300" />
</p>


[![Build Status](https://travis-ci.com/doniseferi/salahtimes.svg?branch=master)](https://travis-ci.com/doniseferi/salahtimes)[![Codacy Badge](https://api.codacy.com/project/badge/Grade/27de579f6fc84188ba0aac2601ec05f0)](https://www.codacy.com/manual/doniseferi/salahtimes?utm_source=github.com&utm_medium=referral&utm_content=doniseferi/salahtimes&utm_campaign=Badge_Grade)![npm](https://img.shields.io/npm/v/salahtimes)
![GitHub](https://img.shields.io/github/license/doniseferi/salahtimes)
![Snyk Vulnerabilities for npm package version](https://img.shields.io/snyk/vulnerabilities/npm/salahtimes)![GitHub last commit](https://img.shields.io/github/last-commit/doniseferi/salahtimes)[![Coverage Status](https://coveralls.io/repos/github/doniseferi/salahtimes/badge.svg?branch=master)](https://coveralls.io/github/doniseferi/salahtimes?branch=master)[![GitHub issues](https://img.shields.io/github/issues/doniseferi/salahtimes)](https://github.com/doniseferi/salahtimes/issues)![GitHub release (latest SemVer)](https://img.shields.io/github/v/release/doniseferi/salahtimes)
* * *
  
A tiny library to get the exact date and time in an [ISO 8601](https://www.wikiwand.com/en/ISO_8601) format for any of the five daily prayers at any location and date.

## Install

```npm install salahtimes```

## Api Reference

* * *

## Functions
  
<dl>
<dt><a href="#getFajrDateTimeUtc">getFajrDateTimeUtc(date, latitude, longitude, islamicConvention, highLatitudeMethod)</a> ⇒ <code>string</code></dt>
<dd><p>Gets the date and time for fajr in UTC expressed in an ISO 8601 format for date instance, latitude, longitude value and optional islamic convention and high laltitude method.</p>
</dd>

<dt><a href="#getDhuhrDateTimeUtc">getDhuhrDateTimeUtc(date, longitude)</a> ⇒ <code>string</code></dt>
<dd><p>Gets the date and time for dhuhr in UTC expressed in an ISO 8601 format for date instance and longitude value.</p>
</dd>

<dt><a href="#getAsrDateTimeUtc">getAsrDateTimeUtc(date, latitude, longitude, madhab)</a> ⇒ <code>string</code></dt>
<dd><p>Gets the date and time for asr in UTC expressed in an ISO 8601 format for date instance, latitude and longitude value and optional madhab value.</p>
</dd>

<dt><a href="#getMaghribDateTimeUtc">getMaghribDateTimeUtc(date, longitude): string;</a> ⇒ <code>string</code></dt>
<dd><p>Gets the date and time for maghrib in UTC expressed in an ISO 8601 format for date instance, latitude and longitude value.</p>
</dd>

<dt><a href="#getIshaaDateTimeUtc">getIshaaDateTimeUtc(date, latitude, longitude, islamicConvention, highLatitudeMethod)</a> ⇒ <code>string</code></dt>
<dd><p>Gets the date and time for ishaa in UTC expressed in an ISO 8601 format for date instance, latitude, longitude value and optional islamic convention and high laltitude method.</p>
</dd>



</dl>

* * *
  
<a name="getFajrDateTimeUtc"></a>

### `Const` getFajrDateTimeUtc

▸ **getFajrDateTimeUtc**(`date`: Date, `latitude`: number, `longitude`: number, `islamicConvention`: [SupportedConventions](#SupportedConventions), `highLatitudeMethod`: [HighLatitudeMethod](#HighLatitudeMethod)): *string*

*Defined in [src/client/adapter.ts:16](https://github.com/doniseferi/salahtimes/blob/ceee6ba/src/client/adapter.ts#L16)*

**Parameters:**

Name | Type | Default |
------ | ------ | ------ |
`date` | Date | - |
`latitude` | number | - |
`longitude` | number | - |
`islamicConvention` | [SupportedConventions](#SupportedConventions) | "MuslimWorldLeague" |
`highLatitudeMethod` | [HighLatitudeMethod](#HighLatitudeMethod) | "AngleBasedMethod" |

**Returns:** *string*

**Example** *(Example usage of getFajrDateTimeUtc)*

```js
// returns '2021-03-01T04:57:11.422Z'
getFajrDateTimeUtc(new Date(2021, 2, 1), 51.522079, -0.191380);
```

**Example** *(Example usage of getFajrDateTimeUtc)*

```js
// returns '2021-03-01T04:47:21.866Z'
getFajrDateTimeUtc(new Date(2021, 2, 1), 51.522079, -0.191380, 'EgyptianGeneralAuthorityOfSurvey');
```

**Example** *(Example usage of getFajrDateTimeUtc)*

```js
* // returns '2021-03-01T04:47:21.866Z'
* getFajrDateTimeUtc(new Date(2021, 2, 1), 51.522079, -0.191380, 'EgyptianGeneralAuthorityOfSurvey'. 'MiddleOfTheNightMethod');
```

* * *

<a name="getDhuhrDateTimeUtc"></a>

### `Const` getDhuhrDateTimeUtc

▸ **getDhuhrDateTimeUtc**(`date`: Date, `longitude`: number): *string*

*Defined in [src/client/adapter.ts:29](https://github.com/doniseferi/salahtimes/blob/8de06c5/src/client/adapter.ts#L29)*

**Parameters:**

Name | Type |
------ | ------ |
`date` | Date |
`longitude` | number |

**Returns:** *string*

**Example** *(Example usage of getDhuhrDateTimeUtc)*

```js
// returns "2025-01-18T12:10:20.853Z"
getDhuhrDateTimeUtc(new Date(2025, 0, 18), -0.010150);
```

* * *

<a name="getAsrDateTimeUtc"></a>

### `Const` getAsrDateTimeUtc

▸ **getAsrDateTimeUtc**(`date`: Date, `latitude`: number, `longitude`: number, `madhab`: [Madhab](#Madhab)): *string*

*Defined in [src/client/adapter.ts:41](https://github.com/doniseferi/salahtimes/blob/ceee6ba/src/client/adapter.ts#L41)*

**Parameters:**

Name | Type | Default |
------ | ------ | ------ |
`date` | Date | - |
`latitude` | number | - |
`longitude` | number | - |
`madhab` | [Madhab](#Madhab) | "Standard" |

**Returns:** *string*

**Example** (Example usage of getAsrDateTimeUtc)
```js
// returns "2021-03-01T15:02:05.884Z"
getAsrDateTimeUtc(new Date(2021, 2, 1), 51.522079, -0.191380);
```

**Example** (Example usage of getAsrDateTimeUtc)

```js
// returns "2021-03-01T15:02:05.884Z"
getAsrDateTimeUtc(new Date(2021, 2, 1), 51.522079, -0.191380, 'Shafii');
```

**Example** (Example usage of getAsrDateTimeUtc)

```js
// returns "2021-03-01T15:45:41.293Z"
getAsrDateTimeUtc(new Date(2021, 2, 1), 51.522079, -0.191380, 'Hanafi');
```

* * *

<a name="getMaghribDateTimeUtc"></a>
<a name="getFajrDateTimeUtc"></a>

### `Const` getMaghribDateTimeUtc

▸ **getMaghribDateTimeUtc**(`date`: Date, `latitude`: number,`longitude`: number): *string*

*Defined in [src/client/adapter.ts:54](https://github.com/doniseferi/salahtimes/blob/8de06c5/src/client/adapter.ts#L54)*

**Parameters:**

Name | Type |
------ | ------ |
`date` | Date |
`latitude` | number |
`longitude` | number |

**Returns:** *string*

**Example** *(Example usage of getMaghribDateTimeUtc)*

```js
// returns "2025-01-18T12:10:20.853Z"
getMaghribDateTimeUtc(new Date(2025, 0, 18), -0.010150);
```

**Example** *(Example usage of getMaghribDateTimeUtc)*

```js
// returns '2037-08-02T17:59:44.502Z'
getMaghribDateTimeUtc(new Date(2037, 7, 2), 42.637610, 21.092160);
```

**Example** *(Example usage of getMaghribDateTimeUtc)*

```js
// returns 'The sun is up all day on 2037-08-01T23:00:00.000Z at latitude: 71.98007 and longitude: 102.47427'
getMaghribDateTimeUtc(new Date(2037, 7, 2), 71.980070, 102.474270);
```

**Example** *(Example usage of getMaghribDateTimeUtc)*

```js
// returns 'The sun is down all day on 2032-01-01T00:00:00.000Z at latitude: 89.525 and longitude: -30.45'
getMaghribDateTimeUtc(new Date(2032, 0, 1), 89.5250, -30.4500);
```

**Example** *(Example usage of getMaghribDateTimeUtc)*

```js
// returns 'The sun is down all day on 2032-01-01T00:00:00.000Z at latitude: 89.525 and longitude: -30.45'
getMaghribDateTimeUtc(new Date(2032, 0, 1), 89.5250, -30.4500);
```

* * *

<a name="getIshaaDateTimeUtc"></a>


### `Const` getIshaaDateTimeUtc

▸ **getIshaaDateTimeUtc**(`date`: Date, `latitude`: number, `longitude`: number, `islamicConvention`: [SupportedConventions](#SupportedConventions), `highLatitudeMethod`: [HighLatitudeMethod](#HighLatitudeMethod)): *string*

*Defined in [src/client/adapter.ts:16](https://github.com/doniseferi/salahtimes/blob/ceee6ba/src/client/adapter.ts#L63)*

**Parameters:**

Name | Type | Default |
------ | ------ | ------ |
`date` | Date | - |
`latitude` | number | - |
`longitude` | number | - |
`islamicConvention` | [SupportedConventions](#SupportedConventions) | "MuslimWorldLeague" |
`highLatitudeMethod` | [HighLatitudeMethod](#HighLatitudeMethod) | "AngleBasedMethod" |

**Returns:** *string*

**Example** *(Example usage of getIshaaDateTimeUtc)*
```js
// returns '2021-03-01T19:22:33.506Z'
getIshaaDateTimeUtc(new Date(2021, 2, 1), 51.522079, -0.191380);
```
**Example** *(Example usage of getIshaaDateTimeUtc)*
```js
// returns '2021-03-01T19:29:04.467Z'
getIshaaDateTimeUtc(new Date(2021, 2, 1), 51.522079, -0.191380, 'UniversityOfIslamicSciencesKarachi');
```
**Example** *(Example usage of getIshaaDateTimeUtc)*
```js
// returns '2021-03-01T19:29:04.467Z'
getIshaaDateTimeUtc(new Date(2021, 2, 1), 51.522079, -0.191380, 'UniversityOfIslamicSciencesKarachi', 'MiddleOfTheNightMethod');
```

* * *

<a name="SupportedConventions"><a/>

###  SupportedConventions

Ƭ **SupportedConventions**: *"MuslimWorldLeague" | "IslamicSocietyOfNorthAmerica" | "EgyptianGeneralAuthorityOfSurvey" | "UmmAlQuraUniversityMekkah" | "UniversityOfIslamicSciencesKarachi" | "InstituteOfGeophysicsUniversityOfTehranOfSurvey" | "ShiaIthnaAshariLevaResearchInstituteQumOfSurvey"*

* * *

<a name="HighLatitudeMethod"><a/>

###  HighLatitudeMethod

Ƭ **HighLatitudeMethod**: *"AngleBasedMethod" | "MiddleOfTheNightMethod" | "OneSeventhMethod"*

* * *

###  Madhab

<a name="Madhab"><a/>

Ƭ **Madhab**: *"Standard" | "Shafii" | "Maliki" | "Hanbali" | "Hanafi"*

* * *


[More indepth implementation documentation](./docs/interfaces/../README.md)

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.1.15](https://github.com/doniseferi/salahtimes/compare/v1.1.14...v1.1.15) (2020-08-15)

### [1.1.14](https://github.com/doniseferi/salahtimes/compare/v1.1.13...v1.1.14) (2020-08-12)

### [1.1.13](https://github.com/doniseferi/salahtimes/compare/v1.1.12...v1.1.13) (2020-08-12)

### [1.1.12](https://github.com/doniseferi/salahtimes/compare/v1.1.11...v1.1.12) (2020-08-12)

### [1.1.10](https://github.com/doniseferi/salahtimes/compare/v1.1.9...v1.1.10) (2020-07-29)

### [1.1.9](https://github.com/doniseferi/salahtimes/compare/v1.1.8...v1.1.9) (2020-07-29)

### [1.1.8](https://github.com/doniseferi/salahtimes/compare/v1.1.7...v1.1.8) (2020-07-29)

### [1.1.7](https://github.com/doniseferi/salahtimes/compare/v1.1.6...v1.1.7) (2020-07-28)

### [1.1.6](https://github.com/doniseferi/salahtimes/compare/v1.1.5...v1.1.6) (2020-07-27)

### [1.1.5](https://github.com/doniseferi/salahtimes/compare/v1.1.4...v1.1.5) (2020-07-26)

### [1.1.4](https://github.com/doniseferi/salahtimes/compare/v1.1.3...v1.1.4) (2020-07-26)


salahtimes.d.ts
// Type definitions for salahtimes
// Project: salahtimes
// Definitions by: Edonis Seferi https://github.com/doniseferi
export as namespace salahtimes;

type Madhab = "Standard" | "Shafii" | "Maliki" | "Hanbali" | "Hanafi";
type HighLatitudeMethod =
  | "AngleBasedMethod"
  | "MiddleOfTheNightMethod"
  | "OneSeventhMethod";
type SupportedConventions =
  | "MuslimWorldLeague"
  | "IslamicSocietyOfNorthAmerica"
  | "EgyptianGeneralAuthorityOfSurvey"
  | "UmmAlQuraUniversityMekkah"
  | "UniversityOfIslamicSciencesKarachi"
  | "InstituteOfGeophysicsUniversityOfTehranOfSurvey"
  | "ShiaIthnaAshariLevaResearchInstituteQumOfSurvey";

/**
 * Gets the date and time for fajr in UTC expressed in an ISO 8601 format for date instance, latitude and longitude value.
 * @example Example usage of getFajrDateTimeUtc
 * // returns '2021-03-01T04:57:11.422Z'
 * getFajrDateTimeUtc(new Date(2021, 2, 1), 51.522079, -0.191380);
 * @example Example usage of getFajrDateTimeUtc
 * // returns '2021-03-01T04:47:21.866Z'
 * getFajrDateTimeUtc(new Date(2021, 2, 1), 51.522079, -0.191380, 'EgyptianGeneralAuthorityOfSurvey');
 * @example Example usage of getFajrDateTimeUtc
 * // returns '2021-03-01T04:47:21.866Z'
 * getFajrDateTimeUtc(new Date(2021, 2, 1), 51.522079, -0.191380, 'EgyptianGeneralAuthorityOfSurvey'. 'MiddleOfTheNightMethod');
 * @param {Date} date - A date instance.
 * @param {number} latitude - A latitude value in the range of -90 to 90.
 * @param {number} longitude - A longitude value in the range of -180 to 180.
 * @param {SupportedConventions} [islamicConvention="MuslimWorldLeague"] - The islamic convention that defines the angle to be used for fajr.
 * @param {HighLatitudeMethod} [highLatitudeMethod="AngleBasedMethod"] - The method to use when the sun does not fall at the angle required for fajr.
 * @returns {string} A string representing date and time for fajr in UTC, expressed in an ISO 8601 format.
 * @function
 */
export function getFajrDateTimeUtc(
  date: Date,
  latitude: number,
  longitude: number,
  islamicConvention?: SupportedConventions,
  highLatitudeMethod?: HighLatitudeMethod
): string;

/**
 * Gets the date and time for dhuhr in UTC expressed in an ISO 8601 format for date instance, latitude and longitude value.
 * @example Example usage of getDhuhrDateTimeUtc
 * // returns "2025-01-18T12:10:20.853Z"
 * getDhuhrDateTimeUtc(new Date(2025, 0, 18), -0.010150);
 * @param {Date} date - A date instance.
 * @param {number} longitude - A longitude value in the range of -180 to 180.
 * @returns {string} A string representing date and time for dhuhr in UTC, expressed in an ISO 8601 format.
 * @function
 */
export function getDhuhrDateTimeUtc(date: Date, longitude: number): string;

/**
 * Gets the date and time for asr in UTC expressed in an ISO 8601 format for date instance, latitude and longitude value.
 * @example Example usage of getAsrDateTimeUtc
 * // returns "2021-03-01T15:02:05.884Z"
 * getAsrDateTimeUtc(new Date(2021, 2, 1), 51.522079, -0.191380);
 * // returns "2021-03-01T15:02:05.884Z"
 * getAsrDateTimeUtc(new Date(2021, 2, 1), 51.522079, -0.191380, 'Shafii');
 * @example Example usage of getAsrDateTimeUtc
 * // returns "2021-03-01T15:45:41.293Z"
 * getAsrDateTimeUtc(new Date(2021, 2, 1), 51.522079, -0.191380, 'Hanafi');
 * @param {Date} date - A date instance.
 * @param {number} latitude - A latitude value in the range of -90 to 90.
 * @param {number} longitude - A longitude value in the range of -180 to 180.
 * @param {Madhab} [madhab="Standard"] - The madhab value for determining the asr elevation angle. Hanafi is the only value that produces a different value in comparison to the other values.
 * @returns {string} A string representing date and time for asr in UTC, expressed in an ISO 8601 format.
 * @function
 */
export function getAsrDateTimeUtc(
  date: Date,
  latitude: number,
  longitude: number,
  madhab?: Madhab
): string;

/**
 * Gets the date and time for maghrib in UTC expressed in an ISO 8601 format for date instance, latitude and longitude value.
 * @example Example usage of getMaghribDateTimeUtc
 * // returns '2037-08-02T17:59:44.502Z'
 * getMaghribDateTimeUtc(new Date(2037, 7, 2), 42.637610, 21.092160)
 * @example Example usage of getMaghribDateTimeUtc
 * // returns 'The sun is up all day on 2037-08-01T23:00:00.000Z at latitude: 71.98007 and longitude: 102.47427'
 * getMaghribDateTimeUtc(new Date(2037, 7, 2), 71.980070, 102.474270)
 * @example Example usage of getMaghribDateTimeUtc
 * // returns 'The sun is down all day on 2032-01-01T00:00:00.000Z at latitude: 89.525 and longitude: -30.45'
 * getMaghribDateTimeUtc(new Date(2032, 0, 1), 89.5250, -30.4500)
 * @example Example usage of getMaghribDateTimeUtc
 * // returns 'The sun is down all day on 2032-01-01T00:00:00.000Z at latitude: 89.525 and longitude: -30.45'
 * getMaghribDateTimeUtc(new Date(2032, 0, 1), 89.5250, -30.4500)
 * @param {Date} date - A date instance.
 * @param {number} latitude - A latitude value in the range of -90 to 90.
 * @param {number} longitude - A longitude value in the range of -180 to 180.
 * @returns {string} A string representing date and time for maghrib in UTC, expressed in an ISO 8601 format.
 * @function
 */
export function getMaghribDateTimeUtc(
  date: Date,
  latitude: number,
  longitude: number
): string;

/**
 * Gets the date and time for fajr in UTC expressed in an ISO 8601 format for date instance, latitude and longitude value.
 * @example Example usage of getIshaaDateTimeUtc
 * // returns '2021-03-01T19:22:33.506Z'
 * getIshaaDateTimeUtc(new Date(2021, 2, 1), 51.522079, -0.191380)
 * @example Example usage of getIshaaDateTimeUtc
 * // returns '2021-03-01T19:29:04.467Z'
 * getIshaaDateTimeUtc(new Date(2021, 2, 1), 51.522079, -0.191380, 'UniversityOfIslamicSciencesKarachi')
 * @example Example usage of getIshaaDateTimeUtc
 * // returns '2021-03-01T19:29:04.467Z'
 * getIshaaDateTimeUtc(new Date(2021, 2, 1), 51.522079, -0.191380, 'UniversityOfIslamicSciencesKarachi', 'MiddleOfTheNightMethod')
 * @param {Date} date - A date instance.
 * @param {number} latitude - A latitude value in the range of -90 to 90.
 * @param {number} longitude - A longitude value in the range of -180 to 180.
 * @param {SupportedConventions} [islamicConvention="MuslimWorldLeague"] - The islamic convention that defines the angle to be used for ishaa.
 * @param {HighLatitudeMethod} [highLatitudeMethod="AngleBasedMethod"] - The method to use when the sun does not fall at the angle required for ishaa.* @returns {string} A string representing date and time for ishaa in UTC, expressed in an ISO 8601 format.
 * @function
 */
export function getIshaaDateTimeUtc(
  date: Date,
  latitude: number,
  longitude: number,
  islamicConvention?: SupportedConventions,
  highLatitudeMethod?: HighLatitudeMethod
): string;
